src/alcazar.rs

```
@@ -1,51 +1,88 @@
-use crate::http_request::HttpRequest;
-use std::net::{SocketAddr, TcpListener};
+use crate::{http_request::HttpRequest, router::Router};
+use std::{
+    io::Write,
+    net::{SocketAddr, TcpListener},
+};
 use tracing::info;
 
 #[derive(Default)]
 pub struct AlcazarBuilder {
     url: Option<SocketAddr>,
+    router: Router,
 }
 
 impl AlcazarBuilder {
     pub fn set_addr(self, url: SocketAddr) -> Self {
-        Self { url: url.into() }
+        Self {
+            url: url.into(),
+            router: self.router,
+        }
+    }
+
+    pub fn set_router(self, router: Router) -> Self {
+        Self {
+            url: self.url,
+            router,
+        }
     }
 
     pub fn start(self) -> Alcazar {
         let listener =
             TcpListener::bind(self.url.unwrap_or_else(|| "0.0.0.0:0".parse().unwrap())).unwrap();
```

Not sure about the correctnes for the `0` port. Why not to use `8000` by default as my other frameworks does?

---

Instead of using `unwrap` we could return an error here instead. It's much more relieble and easier to understand what goes wrong

---

> Not sure about the correctnes for the 0 port. Why not to use 8000 by default as my other frameworks does?

0 means pick one up for me, 8080 sounds good to me as long as clients can specify the 0 port themselves :)

---

I add an `expect` instead of an unwrap, we will have to handle the `Error` and implement `Result` in an other PR.